frontend/src/routes/Video.tsx

```diff
@@ -137,12 +143,90 @@ export const VideoRoute = makeRoute({
     },
 });
 
+/** Video in realm route: `/path/to/realm/v/:<ocid>` */
+export const OpencastVideoRoute = makeRoute({
+    url: ({ realmPath, ocId }: { realmPath: string; ocId: string }) =>
+        `${realmPath === "/" ? "" : realmPath}/v/:${ocId}`,
+    match: url => {
+        const urlPath = url.pathname.replace(/^\/|\/$/g, "");
+        const parts = urlPath.split("/").map(decodeURIComponent);
+        if (parts.length < 2) {
+            return null;
+        }
+        if (parts[parts.length - 2] !== "v") {
+            return null;
+        }
+        let videoId = parts[parts.length - 1];
+        if (!videoId.match(new RegExp(":([^/]+)"))) {
+            return null;
+        } else {
+            videoId = videoId.substring(1);
+        }
+
+        const realmPathParts = parts.slice(0, parts.length - 2);
+        if (!isValidRealmPath(realmPathParts)) {
+            return null;
+        }
+
+        const query = graphql`
+            query VideoPageByOcIdInRealmQuery($id: String!, $realmPath: String!) {
+                ... UserData
+                event: eventByOpencastId(id: $id) {
+                    ... VideoPageEventData
+                    ... on AuthorizedEvent {
+                        isReferencedByRealm(path: $realmPath)
+                    }
+                }
+                realm: realmByPath(path: $realmPath) {
+                    ... VideoPageRealmData
+                    ... NavigationData
+                }
+            }
+        `;
+
+        const realmPath = "/" + realmPathParts.join("/");
+        const queryRef = loadQuery<VideoPageByOcIdInRealmQuery>(query, {
+            id: videoId,
+            realmPath,
+        });
+
+        return {
+            render: () => <RootLoader
+                {... { query, queryRef }}
+                nav={data => data.realm ? <Nav fragRef={data.realm} /> : []}
+                render={({ event, realm }) => {
+                    if (!event) {
+                        return <NotFound kind="video" />;
+                    }
+
+                    if (!realm || !event.isReferencedByRealm) {
+                        return <ForwardToDirectOcRoute ocID={videoId} />;
+                    }
+
+                    return <VideoPage
+                        eventRef={event}
+                        realmRef={realm}
+                        basePath={realmPath.replace(/\/$/u, "") + "/v"}
+                    />;
+                }}
+            />,
+            dispose: () => queryRef.dispose(),
+        };
+    },
```

Comment:
This has lots of duplication with the other route. Can you extract at least some of it into a helper function somehow?